### PR TITLE
fix: Blame index update

### PR DESCRIPTION
**Pull request by kevinssgh**
# Description

* Updated format of blame index: chainId-nonce-msgdigest-height
* Added query for blame records by chainId and nonce
* Added cmd for new query

Closes: <PD-XXXX>

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Include instructions and any relevant details so others can reproduce. 

- [ ] Tested CCTX in localnet
- [ ] Tested in development environment
- [ ] Go unit tests
- [ ] Go integration tests
- [ ] Tested via GitHub Actions 

# Checklist:

- [ ] I have added unit tests that prove my fix feature works

### CODE REVIEW

File: proto/observer/query.proto [hunk 1]
```diff
@@ -211,6 +216,15 @@ message QueryAllBlameRecordsResponse {
   repeated Blame blame_info = 1;
 }
 
+message QueryBlameByChainAndNonceRequest {
+  uint32 chain_id = 1;
```
Comment (ws4charlie) [inline proto/observer/query.proto:220]: can we unify chain_id type as `int64` like other in other `query.proto` files? The `QueryProveRequest` one was was already updated in one of my pending PRs.

**Review by ws4charlie — APPROVED** [review]
LGTM

File: x/observer/keeper/blame.go [hunk 2]
```diff
@@ -40,6 +41,21 @@ func (k Keeper) GetAllBlame(ctx sdk.Context) (BlameRecords []*types.Blame, found
 	return
 }
 
+func (k Keeper) GetBlameByChainAndNonce(ctx sdk.Context, chainID int64, nonce int64) (BlameRecords []*types.Blame, found bool) {
+	store := prefix.NewStore(ctx.KVStore(k.storeKey), types.KeyPrefix(types.BlameKey))
+	blamePrefix := fmt.Sprintf("%d-%d", chainID, nonce)
```
Comment (lumtis) [inline x/observer/keeper/blame.go:46]: I looked into the code, I see that ZetaClient index the blame index with:
```
func digestReq(request keygen.Request) (string, error) {
	bytes, err := json.Marshal(request)
	if err != nil {
		return "", err
	}

	hasher := sha3.NewLegacyKeccak256()
	hasher.Write(bytes)
	digest := hex.EncodeToString(hasher.Sum(nil))

	return digest, nil
}
```
Therefore, I'm not sure to understand how they would be indexed with `[chainID]-[nonce]`

**Review by lumtis — COMMENTED** [review]
We should use plural blames since the query returns a list of blames

File: x/observer/keeper/blame.go (hunk 2 above)
Comment (kevinssgh) [reply x/observer/keeper/blame.go:46]: This particular digest is for keygen. There won't be any chain or nonce associated with it, I can however add a different prefix to this like "keygen-" to differentiate it. 

The change was really intended for keysign failures.

File: x/observer/keeper/blame_test.go [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+	index := fmt.Sprintf("%d-%d-%s-%d", chainId, nonce, digest, 123)
+
+	keeper.SetBlame(ctx, &types.Blame{
+		Index:         index,
+		FailureReason: "failed to join party",
+		Nodes:         nil,
+	})
+
+	blameRecords, found := keeper.GetBlame(ctx, index)
+	assert.True(t, found)
```
Comment (lumtis) [inline x/observer/keeper/blame_test.go:27]: I think we can use `require` insert of `assert` everywhere

File: zetaclient/tss_signer.go [hunk 4]
```diff
@@ -112,7 +112,7 @@ func (tss *TSS) Sign(digest []byte, height uint64, chain *common.Chain, optional
 		log.Warn().Msgf("keysign status FAIL posting blame to core, blaming node(s): %#v", ksRes.Blame.BlameNodes)
 
 		digest := hex.EncodeToString(digest)
-		index := fmt.Sprintf("%s-%d", digest, height)
+		index := fmt.Sprintf("%d-%d-%s-%d", chain.ChainId, nonce, digest, height)
```
Comment (lumtis) [inline zetaclient/tss_signer.go:115]: We generally have the code to calculate prefixes in `keys.go`, `x/observer/types/keys.go`
Maybe we could implement:
```
GetBlameIndex(chainID, nonce, digest, height)
GetBlamePrefix(chainID, nonce)
```
to use it here and in the query, so the low level aspect of calculating the prefix index is abstracted

File: zetaclient/tss_signer.go [hunk 5]
```diff
@@ -5,6 +5,7 @@ import (
 	"encoding/base64"
 	"encoding/hex"
 	"fmt"
+	observerTypes "github.com/zeta-chain/zetacore/x/observer/types"
```
Comment (lumtis) [inline zetaclient/tss_signer.go:8]:
```suggestion
	observertypes "github.com/zeta-chain/zetacore/x/observer/types"
```

Comment (lumtis) [inline zetaclient/tss_signer.go:8]: We should run `make generate`

**Review by lumtis — COMMENTED** [review]
Minor comments, looks good

**Review by lumtis — APPROVED** [review]

**Pull request merged**